### PR TITLE
[SPARK-48566][PYTHON] Fix bug where partition indices are incorrect when UDTF analyze() uses both select and partitionColumns

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR fixes a bug that resulted in an internal error with some combination of the Python UDTF "select" and "partitionBy" options of the "analyze" method.

Specifically, this logic in `Analyzer.scala` was wrong because it did not update the usage of `partitioningExpressionIndexes` to take the "select" expressions into account when they were introduced in https://github.com/apache/spark/pull/45007:

```
val tvfWithTableColumnIndexes = tvf match {
  case g @ Generate(pyudtf: PythonUDTF, _, _, _, _, _)
      if tableArgs.head._1.partitioningExpressionIndexes.nonEmpty =>
    //////////////////////////////////////////////////////////////////////////////
    // The bug is here: the 'partitioningExpressionIndexes' are not valid
    // if the UDTF "select" expressions are non-empty, since that prompts
    // us to add a new projection (of a possibly different number of
    // expressions) to evaluate them.
    //////////////////////////////////////////////////////////////////////////////
    val partitionColumnIndexes =
      PythonUDTFPartitionColumnIndexes(tableArgs.head._1.partitioningExpressionIndexes)
    g.copy(generator = pyudtf.copy(
      pythonUDTFPartitionColumnIndexes = Some(partitionColumnIndexes)))
  case _ => tvf
}
```

To reproduce:

```
from pyspark.sql.functions import (
    AnalyzeArgument,
    AnalyzeResult,
    PartitioningColumn,
    SelectedColumn,
    udtf
)

from pyspark.sql.types import (
    DoubleType,
    StringType,
    StructType,
)

@udtf
class TestTvf:
    @staticmethod
    def analyze(observed: AnalyzeArgument) -> AnalyzeResult:
        out_schema = StructType()
        out_schema.add("partition_col", StringType())
        out_schema.add("double_col", DoubleType())

        return AnalyzeResult(
            schema=out_schema,
            partitionBy=[PartitioningColumn("partition_col")],
            select=[
                SelectedColumn("partition_col"),
                SelectedColumn("double_col"),
            ],
        )

    def eval(self, *args, **kwargs):
        pass

    def terminate(self):
        for _ in range(10):
            yield {
                "partition_col": None,
                "double_col": 1.0,
            }


spark.udtf.register("serialize_test", TestTvf) 

# Fails
(
    spark
    .sql(
        """
        SELECT * FROM serialize_test(
            TABLE(
                SELECT
                    5 AS unused_col,
                    'hi' AS partition_col,
                    1.0 AS double_col
                
                UNION ALL

                SELECT
                    4 AS unused_col,
                    'hi' AS partition_col,
                    1.0 AS double_col
            )
        )
        """
    )
    .toPandas()
)
```

### Why are the changes needed?

The above query returned internal errors before, but works now.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Additional golden file coverage

### Was this patch authored or co-authored using generative AI tooling?

Some light GitHub copilot usage

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/FunctionTableSubqueryArgumentExpression.scala [hunk 1]
```diff
@@ -165,20 +165,18 @@ case class FunctionTableSubqueryArgumentExpression(
    * when the partition has changed.
    */
   lazy val partitioningExpressionIndexes: Seq[Int] = {
-    val extraPartitionByExpressionsToIndexes: Map[Expression, Int] =
-      extraProjectedPartitioningExpressions.map(_.child).zipWithIndex.toMap
-    partitionByExpressions.map { e =>
-      subqueryOutputs.getOrElse(e, extraPartitionByExpressionsToIndexes(e) + plan.output.length)
+    (0 until extraProjectedPartitioningExpressions.length).map { i =>
+      if (selectedInputExpressions.nonEmpty) {
+        selectedInputExpressions.length + i
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/FunctionTableSubqueryArgumentExpression.scala:170]: Why is this related to `selectedInputExpressions`?

Comment (ueshin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/FunctionTableSubqueryArgumentExpression.scala:170]: This will break
https://github.com/apache/spark/blob/51188ed56620c17841dac60f7f2a9330a1087d6e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/FunctionTableSubqueryArgumentExpression.scala#L124
if `selectedInputExpressions.length` and `plan.output.length` are different.

Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/FunctionTableSubqueryArgumentExpression.scala]: Good catch. This is related to `selectedInputExpressions` because if they are set, we want to use different indexes here [1]. I switched to update that location instead, which should no longer break the location you mentioned.

[1] https://github.com/apache/spark/blob/201df0d7ac81f6bd5c39f513b0a06cb659dc9a3f/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L2211

**Review by dtenedor — COMMENTED** [review]
Thanks @ueshin, responded to you comment, please take another look!

**Review by ueshin — COMMENTED** [review]
cc @allisonwang-db

**Review by ueshin — APPROVED** [review]

**Comment by dtenedor** [comment]
👍 @ueshin I updated the PR description to clarify the exact bug fix more. All CI tests are passing, please feel free to merge at your convenience :)

**Review by allisonwang-db — APPROVED** [review]

**Comment by ueshin** [comment]
Thanks! merging to master.
